File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    genres_cnt = {}
    sorted_genres = {}
    for idx, data in enumerate(zip(genres, plays)):
        genre = data[0]
        play = data[1]

        if genre not in genres_cnt.keys():
            genres_cnt[genre] = 0

        if genre not in sorted_genres.keys():
            sorted_genres[genre] = []

        genres_cnt[genre] += play
        sorted_genres[genre].append([idx, play])
        idx += 1

    orders = sorted(genres_cnt, key=lambda x: -genres_cnt[x])
    for genre in orders:
        sorted_genres[genre].sort(key=lambda x: -x[1])

        for i in range(len(sorted_genres[genre])):
            if i > 1:
                break
            idx = sorted_genres[genre][i][0]
            answer.append(idx)

    return answer

File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.kt
class Solution {
    fun solution(genres: Array<String>, plays: IntArray): IntArray {

        val genreTotalPlays = mutableMapOf<String, Int>()
        val genreSongs = mutableMapOf<String, MutableList<Pair<Int, Int>>>()

        for (i in genres.indices) {
            val genre = genres[i]
            val playCount = plays[i]

            genreTotalPlays[genre] = genreTotalPlays.getOrDefault(genre, 0) + playCount

            if (!genreSongs.containsKey(genre)) {
                genreSongs[genre] = mutableListOf(Pair(i, playCount))
            } else {
                genreSongs[genre]?.add(Pair(i, playCount))
            }
        }

        val sortedGenres = genreTotalPlays.entries.sortedByDescending { it.value }.map { it.key }

        val bestAlbum = mutableListOf<Int>()

        for (genre in sortedGenres) {
            val sortedSongs = genreSongs[genre]?.sortedWith(compareBy({ -it.second }, { it.first }))

            if (sortedSongs != null) {
                for (i in sortedSongs.indices) {
                    if (i < 2) {
                        bestAlbum.add(sortedSongs[i].first)
                    } else {
                        break
                    }
                }
            }
        }

        return bestAlbum.toIntArray()
    }
}